File: memind-integrations/openclaw/src/state.ts
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

import { mkdir, readFile, stat, writeFile } from 'node:fs/promises'
import path from 'node:path'

import type { AgentTimelineEvent } from './types.js'

type StateFile = {
  version: 1
  submitted: string[]
}

type AgentTimelineStateFile = {
  version: 1
  events: AgentTimelineEvent[]
}

export class SubmittedStateStore {
  constructor(
    private readonly root: string,
    private readonly options: { maxAgeDays: number },
  ) {}

  async filterUnsubmitted(sessionKey: string, fingerprints: string[]): Promise<string[]> {
    const state = await this.read(sessionKey)
    const submitted = new Set(state.submitted)
    return fingerprints.filter((fingerprint) => !submitted.has(fingerprint))
  }

  async markSubmitted(sessionKey: string, fingerprints: string[]): Promise<void> {
    const state = await this.read(sessionKey)
    const next = new Set(state.submitted)
    fingerprints.forEach((fingerprint) => next.add(fingerprint))
    await this.write(sessionKey, { version: 1, submitted: [...next].sort() })
  }

  private async read(sessionKey: string): Promise<StateFile> {
    try {
      const statePath = this.pathFor(sessionKey)
      const info = await stat(statePath)
      const ageDays = (Date.now() - info.mtimeMs) / 86_400_000
      if (ageDays > this.options.maxAgeDays) return { version: 1, submitted: [] }
      const raw = await readFile(statePath, 'utf8')
      const parsed = JSON.parse(raw) as StateFile
      if (parsed.version === 1 && Array.isArray(parsed.submitted)) return parsed
    } catch {
      return { version: 1, submitted: [] }
    }
    return { version: 1, submitted: [] }
  }

  private async write(sessionKey: string, state: StateFile): Promise<void> {
    await mkdir(this.root, { recursive: true })
    await writeFile(this.pathFor(sessionKey), JSON.stringify(state, null, 2), 'utf8')
  }

  private pathFor(sessionKey: string): string {
    return path.join(this.root, `${safeKey(sessionKey)}.json`)
  }
}

export class AgentTimelineStateStore {
  constructor(
    private readonly root: string,
    private readonly options: { maxAgeDays: number; maxEvents: number },
  ) {}

  async appendAgentEvent(sessionKey: string, event: AgentTimelineEvent): Promise<void> {
    if (!event.eventId) return
    const state = await this.read(sessionKey)
    if (state.events.some((candidate) => candidate.eventId === event.eventId)) return
    const events = [...state.events, event]
      .sort(compareEvents)
      .slice(Math.max(0, state.events.length + 1 - this.options.maxEvents))
    await this.write(sessionKey, { version: 1, events })
  }

  async listAgentEvents(sessionKey: string): Promise<AgentTimelineEvent[]> {
    return (await this.read(sessionKey)).events.sort(compareEvents)
  }

  async clearAgentEvents(sessionKey: string, eventIds?: string[]): Promise<void> {
    if (!eventIds || eventIds.length === 0) {
      await this.write(sessionKey, { version: 1, events: [] })
      return
    }
    const submitted = new Set(eventIds)
    const state = await this.read(sessionKey)
    await this.write(sessionKey, {
      version: 1,
      events: state.events.filter((event) => !event.eventId || !submitted.has(event.eventId)),
    })
  }

  async nextAgentEventSeq(sessionKey: string): Promise<number> {
    const events = await this.listAgentEvents(sessionKey)
    const maxSeq = events.reduce((max, event) => Math.max(max, event.seq ?? 0), 0)
    return maxSeq + 1
  }

  private async read(sessionKey: string): Promise<AgentTimelineStateFile> {
    try {
      const statePath = this.pathFor(sessionKey)
      const info = await stat(statePath)
      const ageDays = (Date.now() - info.mtimeMs) / 86_400_000
      if (ageDays > this.options.maxAgeDays) return { version: 1, events: [] }
      const raw = await readFile(statePath, 'utf8')
      const parsed = JSON.parse(raw) as AgentTimelineStateFile
      if (parsed.version === 1 && Array.isArray(parsed.events)) {
        return {
          version: 1,
          events: parsed.events
            .filter((event): event is AgentTimelineEvent => Boolean(event?.eventId))
            .sort(compareEvents),
        }
      }
    } catch {
      return { version: 1, events: [] }
    }
    return { version: 1, events: [] }
  }

  private async write(sessionKey: string, state: AgentTimelineStateFile): Promise<void> {
    await mkdir(this.root, { recursive: true })
    await writeFile(this.pathFor(sessionKey), JSON.stringify(state, null, 2), 'utf8')
  }

  private pathFor(sessionKey: string): string {
    return path.join(this.root, `${safeKey(sessionKey)}.json`)
  }
}

export function safeKey(value: string): string {
  return value.replace(/[^a-zA-Z0-9_.-]+/g, '_') || 'default'
}

function compareEvents(left: AgentTimelineEvent, right: AgentTimelineEvent): number {
  const seq = (left.seq ?? Number.MAX_SAFE_INTEGER) - (right.seq ?? Number.MAX_SAFE_INTEGER)
  if (seq !== 0) return seq
  return String(left.eventId ?? '').localeCompare(String(right.eventId ?? ''))
}

File: memind-integrations/hermes/memind_hermes/state.py
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

import json
import re
from pathlib import Path


class AgentTimelineState:
    def __init__(self, root, max_events=500):
        self.root = Path(root)
        self.max_events = max_events

    def append(self, session_id, event):
        if not event or not event.get("eventId"):
            return
        state = self._read(session_id)
        if any(existing.get("eventId") == event["eventId"] for existing in state["events"]):
            return
        events = sorted([*state["events"], event], key=_event_key)
        if len(events) > self.max_events:
            events = events[-self.max_events :]
        self._write(session_id, {"version": 1, "events": events})

    def list(self, session_id):
        return sorted(self._read(session_id)["events"], key=_event_key)

    def clear(self, session_id, event_ids=None):
        if not event_ids:
            self._write(session_id, {"version": 1, "events": []})
            return
        submitted = set(event_ids)
        events = [event for event in self.list(session_id) if event.get("eventId") not in submitted]
        self._write(session_id, {"version": 1, "events": events})

    def next_seq(self, session_id):
        events = self.list(session_id)
        if not events:
            return 1
        return max(int(event.get("seq") or 0) for event in events) + 1

    def _read(self, session_id):
        try:
            parsed = json.loads(self._path(session_id).read_text())
            if parsed.get("version") == 1 and isinstance(parsed.get("events"), list):
                return {"version": 1, "events": [event for event in parsed["events"] if event.get("eventId")]}
        except Exception:
            pass
        return {"version": 1, "events": []}

    def _write(self, session_id, state):
        self.root.mkdir(parents=True, exist_ok=True)
        self._path(session_id).write_text(json.dumps(state, indent=2, default=str) + "\n")

    def _path(self, session_id):
        return self.root / f"{_safe_key(session_id)}.json"


def _safe_key(value):
    return re.sub(r"[^A-Za-z0-9_.-]+", "_", str(value or "default")) or "default"


def _event_key(event):
    return (int(event.get("seq") or 0), str(event.get("eventId") or ""))
